text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+owned or borrowed.
+
+Owned objects require the recipient to free them when done with them. crABI 1.0
+does not specify the precise mechanism for making a `free` function for the
+allocator available to the recipient. It is the responsibility of the API
+designer to document how to free owned values, whether by guaranteeing which
+allocator allocated them (e.g. the system allocator), or by exporting an
+appropriate `free` function that passes ownership back.
+
+Lifetimes are incredibly valuable for checking the correctness of pointer
+usage, but as lifetimes only exist at compile time, they are not represented in
+the ABI. It's the responsibility of a function declaration to define lifetimes
+correctly, or to document them in the case of an implementation that doesn't
+model lifetimes in its type system.
+
+The crABI ABI also does not encode whether a pointer is mutable or not; both
+are passed equivalently. It's the responsibility of API documentation (or API
+documentation generators) to reflect whether a pointer permits mutation. Any
+language supporting a distinction between mutable and immutable pointers should
+reflect this when declaring a crABI type or function.
+
+# Types
+[types]: #types
+
+For each type, crABI specifies the translation into C types, either directly or
+via other crABI-supported types. Using the type in a struct, union, enum field,
+function parameter, or function return value, results in a C-compatible
+function equivalent to using the specified translation to C types in place of
+the crABI-supported type. This includes the size and alignment of the
+underlying C types.
+
+Note that C supports passing and returning structures by value, and the ABIs of
+of major targets support passing and returning structures containing multiple
+fields in a manner as efficient as passing multiple parameters, such as by
+passing or returning multiple struct fields via multiple registers. Thus, crABI
+often uses representations involving structures passed or returned by value.
+This has the net effect of grouping these together into multiple types, without
+introducing unnecessary indirection.
+
+Some crABI representations use an underlying C type but do not use the full
+value space of that type. In that case, passing or returning values outside the
+valid range invokes undefined behavior. Programs may assume that no values
+outside that range are passed or returned, and in particular the compiler may
+generate code that does not check this assumption, or may optionally include
+validation assertions when debugging.
+
+## `char` - Unicode character type
+
+crABI supports Unicode characters, which in Rust use the `char` type. This
+translates to the C `uint32_t` type. This type will never contain a value
+larger than 0x10FFFF, or a value in the range 0xD800 to 0xDFFF inclusive.
+
+Note that there is no special handling for an array of values of this type,
+which is not equivalent to a string (unless using a UCS-4 encoding).
```

Comment:
It's not equivalent to either a NUL-terminated or a counted UCS-4 encoding either.
```suggestion
which is not equivalent to a string.
```